File: integrasjon/src/main/java/no/nav/melosys/integrasjon/inngangsvilkar/InngangsvilkarConfig.java
package no.nav.melosys.integrasjon.inngangsvilkar;

import io.getunleash.Unleash;
import no.nav.melosys.integrasjon.felles.WebClientConfig;
import no.nav.melosys.integrasjon.felles.mdc.CorrelationIdOutgoingFilter;
import no.nav.melosys.integrasjon.felles.mdc.CorrelationIdOutgoingInterceptor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.DefaultUriBuilderFactory;

import static no.nav.melosys.featuretoggle.ToggleName.MELOSYS_INNGANGSVILKAR_CONSUMER_BRUK_WEBCLIENT;

@Configuration
public class InngangsvilkarConfig implements WebClientConfig {

    @Bean
    @Qualifier("inngangsvilkår")
    public RestTemplate inngangsVilkaarRestTemplate(@Value("${Inngangsvilkaar.url}") String url,
                                                    RestTemplateBuilder restTemplateBuilder,
                                                    CorrelationIdOutgoingInterceptor correlationIdOutgoingInterceptor) {
        return restTemplateBuilder
            .uriTemplateHandler(new DefaultUriBuilderFactory(url))
            .interceptors(correlationIdOutgoingInterceptor)
            .build();
    }

    @Bean
    public InngangsvilkaarConsumer inngangsvilkaarConsumer(
        @Qualifier("inngangsvilkår") RestTemplate restTemplate,
        @Value("${Inngangsvilkaar.url}") String url,
        WebClient.Builder webclientBuilder,
        CorrelationIdOutgoingFilter correlationIdOutgoingFilter,
        Unleash unleash
    ) {

        if (unleash.isEnabled(MELOSYS_INNGANGSVILKAR_CONSUMER_BRUK_WEBCLIENT)) {
            return new InngangsVilkarConsumerWebclientImpl(
                webclientBuilder
                    .baseUrl(url)
                    .filter(correlationIdOutgoingFilter)
                    .filter(errorFilter("Kall mot inngangsvilkår feilet."))
                    .build()
            );
        }

        return new InngangsvilkaarConsumerImpl(restTemplate);
    }
}

File: integrasjon/src/main/kotlin/no/nav/melosys/integrasjon/inngangsvilkar/InngangsvilkarConfig.kt
package no.nav.melosys.integrasjon.inngangsvilkar

import java.util.Collections
import no.nav.melosys.integrasjon.felles.WebClientConfig
import no.nav.melosys.integrasjon.felles.mdc.CorrelationIdOutgoingFilter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.client.WebClient

@Configuration
class InngangsvilkarConfig : WebClientConfig {

    @Bean
    fun inngangsvilkaarWebClient(
        @Value("\${Inngangsvilkaar.url}") url: String,
        webclientBuilder: WebClient.Builder,
        correlationIdOutgoingFilter: CorrelationIdOutgoingFilter,
    ): WebClient {

        return webclientBuilder
            .baseUrl(url)
            .defaultHeaders { httpHeaders ->
                httpHeaders.accept = Collections.singletonList(MediaType.APPLICATION_JSON)
                httpHeaders.contentType = MediaType.APPLICATION_JSON
            }
            .filter(correlationIdOutgoingFilter)
            .filter(errorFilter("Kall mot inngangsvilkår feilet."))
            .build()
    }
}
